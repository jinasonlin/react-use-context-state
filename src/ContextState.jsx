import React, { createContext, useContext, useReducer, useEffect, useCallback } from 'react';

const ContextStateDefaultVaule = [{}, () => { console.warn('use ContextStateProvider'); }];
const ContextState = createContext(ContextStateDefaultVaule);

export const useContextState = () => useContext(ContextState);

export const useGlobalMember = (key, initialState) => {
  const [state, dispatch] = useContext(ContextState);
  const setState = useCallback((value) => {
    dispatch({ [key]: value });
  }, [key, dispatch]);

  useEffect(() => {
    if (!(key in state)) {
      if (typeof initialState === 'function') {
        if (initialState.length >= 1) {
          initialState(setState);
        } else {
          setState(initialState());
        }
      } else if (initialState) {
        setState(initialState);
      }
    }
  }, [key, dispatch]);

  return [state[key], setState];
};

export const useGlobalMemberReady = (...keys) => {
  const [state] = useContext(ContextState);
  const isReady = keys.every(k => k in state);

  return isReady;
};

export const ContextStateProvider = ({ children }) => {
  const [state, dispatch] = useReducer((s, n) => ({ ...s, ...n }), ContextStateDefaultVaule[0]);

  return (
    <ContextState.Provider value={[state, dispatch]}>
      {children}
    </ContextState.Provider>
  );
};

export const ContextStateConsumer = ({ children }) => {
  return (
    <ContextState.Consumer>
      {([state, dispatch]) => { return typeof children === 'function' ? children(state, dispatch) : children; }}
    </ContextState.Consumer>
  );
};

export default useContextState;
