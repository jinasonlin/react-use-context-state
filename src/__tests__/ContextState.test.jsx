import React, { useEffect } from 'react';
import renderer from 'react-test-renderer';
import { useContextState, ContextStateProvider, ContextStateConsumer } from '../ContextState';

describe('ContextState', () => {
  test('renders provider', () => {
    const tree = renderer
      .create(
        <ContextStateProvider>
          <div>hello world</div>
        </ContextStateProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders consumer', () => {
    const tree = renderer
      .create(
        <ContextStateProvider>
          <ContextStateConsumer>
            <div>hello world</div>
          </ContextStateConsumer>
        </ContextStateProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders consumer - function children', () => {
    const tree = renderer
      .create(
        <ContextStateProvider>
          <ContextStateConsumer>
            {() => <div>hello world</div>}
          </ContextStateConsumer>
        </ContextStateProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('useContextState', () => {
  it('setState in useEffect', () => {
    const Say = () => {
      const [state, setState] = useContextState();
      const { say } = state;

      useEffect(() => {
        setState({
          say: 'Hello',
        });
      }, []);

      return <div>Say: {say}</div>;
    };

    const SayOther = () => {
      const [state] = useContextState();
      const { say } = state;

      return <div id="SayOther">SayOther: {say}</div>;
    };

    const App = (
      <ContextStateProvider>
        <Say />
        <SayOther />
      </ContextStateProvider>
    );

    const testRenderer = renderer
      .create(App);
    const testInstance = testRenderer.root;

    // expect(testInstance.findByType(SayOther).findByType('div').children).toEqual(['SayOther: ']);
    expect(testInstance.findByProps({ id: 'SayOther' }).children).toEqual(['SayOther: ']);

    // render after setState
    testRenderer.update(App);

    expect(testInstance.findByProps({ id: 'SayOther' }).children).toEqual(['SayOther: ', 'Hello']);

    const tree = testRenderer.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
