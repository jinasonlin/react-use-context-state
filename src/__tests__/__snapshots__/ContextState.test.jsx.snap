// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextState renders consumer - function children 1`] = `
<div>
  hello world
</div>
`;

exports[`ContextState renders consumer 1`] = `
<div>
  hello world
</div>
`;

exports[`ContextState renders provider 1`] = `
<div>
  hello world
</div>
`;

exports[`useContextState setState in useEffect 1`] = `
Array [
  <div>
    Say: 
    Hello
  </div>,
  <div
    id="SayOther"
  >
    SayOther: 
    Hello
  </div>,
]
`;
